scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1131,6 +1133,47 @@ class MappablePipeJoinEnrichment[T](pipe: TypedPipe[T]) {
 }
 
 /**
+ * This class is to enrich the forceToDisk operation on TypedPipes and allow
+ * us to potentially optimize these operations when they're chained together
+ * redundantly.
+ */
+case class ForceToDiskTypedPipe[T](typedPipe: TypedPipe[T]) extends TypedPipe[T] {
+  import Dsl._
+
+  override def forceToDisk: TypedPipe[T] = this
+
+  override def asPipe[U >: T](fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
+    val pipe = typedPipe.toPipe[U](fieldNames)(flowDef, mode, setter)
+    pipe.forceToDisk
+  }
+
+  override def cross[U](tiny: TypedPipe[U]): TypedPipe[(T, U)] =
+    forceBefore(typedPipe).cross(tiny)
+
+  override def flatMap[U](f: T => TraversableOnce[U]): TypedPipe[U] =
+    forceBefore(typedPipe).flatMap(f)
```

Comment:
yeah, I think these are two issues the users care about:
1) as you say: `t.forceToDisk.forceToDisk == t.forceToDisk`, which is: make `forceToDisk` idempotent. Nice to have, not a huge deal.

2) make sure all `hashJoin` calls happen with a right that has been forced (maybe except the case of a flatMap). But don't force when it is not needed.

Is that right? To address 2), we don't need the new `ForceToDiskTypedPipe` type, right? I think 2) is the main feature. We can come back to the others.